### PR TITLE
Define and activate release v12.1.1 for AWS

### PR DESCRIPTION
**Pull request by glitchcrab**
This is a patch release to

- pull in the updated kiam-app which should have been included in 12.1.0
- Fix the change detection in aws-operator to ensure an upgrade to this release will work
- Fix a problem with network policies (https://github.com/giantswarm/giantswarm/issues/12617)

### CODE REVIEW

File: aws/v12.1.1/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# :zap: Giant Swarm Release v12.1.1 for AWS :zap:
+
+<< Add description here >>
```
Comment (sslavic) [inline aws/v12.1.1/README.md:3]: Is this some some kind of devctl sorcery?

Comment (glitchcrab) [reply aws/v12.1.1/README.md:3]: goddamnit i forgot to fill out the readme. yes this was generated with devctl

File: aws/v12.1.1/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+### cert-manager [2.0.2](https://github.com/giantswarm/cert-manager-app/releases/tag/v2.0.2)
+
+#### Changed
+- Install CRDs during chart installation only. ([#46](https://github.com/giantswarm/cert-manager-app/pull/46))
+
+
+
+### kiam [1.3.2](https://github.com/giantswarm/kiam-app/releases/tag/v1.3.2)
+
+Not found
```
Comment (sslavic) [inline aws/v12.1.1/README.md:17]: Needs updating to 1.4.0 and changelog entry

Comment (glitchcrab) [reply aws/v12.1.1/README.md]: already done ;)

File: aws/v12.1.1/README.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# :zap: Giant Swarm Release v12.1.1 for AWS :zap:
+
+**If you are upgrading from 12.0.0, upgrading to this release will not roll your nodes. It will only update the apps.**
+
+This release upgrades Cert Manager from upstream v0.9.0 to v0.15.2.
```
Comment (sslavic) [inline aws/v12.1.1/README.md:5]:
```suggestion
This release upgrades kiam App from 1.3.1 to v1.4.0, to align with Cert Manager v0.15.2 API changes.
```

File: aws/v12.1.1/README.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+It is part of a larger effort to maximize reliability of upgrades, by keeping 3rd party components within 30 days of upstream releases.
+
+**This upgrade will stop Cert Manager from reconciling your existing resources** due to changes in its API - manual intervention is required to update affected resources. While the **negative impact to your workloads is low-to-none**, to minimize disruption, **we recommend discussing this upgrade with your Solution Engineer first**.
```
Comment (sslavic) [inline aws/v12.1.1/README.md:9]: This note is relevant for users upgrading from pre 12.0.0 versions. Not sure should it be repeated in 12.0.1, and if so, maybe in some modified form.

**Comment by cokiengchiara** [comment]
> This is a patch release to pull in the updated kiam-app which should have been included in 12.1.0

@glitchcrab should we hand this release to Firecracker?

File: aws/v12.1.1/README.md (hunk 4 above)
Comment (cokiengchiara) [reply aws/v12.1.1/README.md:9]: @marians just announced that there is no active AWS v12 release, right? So for now, this statement will apply to all? :)

Comment (sslavic) [reply aws/v12.1.1/README.md:9]: Releases that are not active, are not necessarily removed/archived.

Customers upgrading from say 11.5.0 to 12.1.1 are expected to read all the release notes in between, even deprecated ones. Just because they should, doesn't mean they will. To increase likelihood not to have such important information missed, we sometimes repeat notes across multiple patch releases.

Customer could already have cluster upgraded to 12.1.0 and have sorted out the cert-manager upgrade in the process, so this note would not apply to them.
But, note would apply say to someone upgrading from 12.0.0 to 12.1.1. They could be not aware of (miss to read) release notes in 12.1.0 and then be surprised. Chance is of customer missing being aware of 12.1.0 and it's release notes is higher due that release being very shortly active.

**Comment by glitchcrab** [comment]
> > This is a patch release to pull in the updated kiam-app which should have been included in 12.1.0
> 
> @glitchcrab should we hand this release to Firecracker?

makes sense to me - my part in this is pretty minor :+1:

File: aws/v12.1.1/release.yaml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  - name: cluster-operator
+    releaseOperatorDeploy: true
+    version: 2.3.2
+  - name: containerlinux
+    version: 2512.2.1
+  - name: etcd
+    version: 3.4.10
+  - name: kubernetes
+    version: 1.17.9
+  date: "2020-08-06T12:00:00Z"
```
Comment (njuettner) [inline aws/v12.1.1/release.yaml:62]:
```suggestion
  date: "2020-08-12T12:00:00Z"
```

**Comment by njuettner** [comment]
Added the information about the aws-operator and updated timestamp for tomorrow to roll it out (12th Aug). Plan is to release it tomorrow because I'd like to let my team to have another look on this release 🙂

File: aws/v12.1.1/release.yaml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  - name: cluster-operator
+    releaseOperatorDeploy: true
+    version: 2.3.2
+  - name: containerlinux
+    version: 2512.2.1
+  - name: etcd
+    version: 3.4.10
+  - name: kubernetes
+    version: 1.17.9
+  date: "2020-08-12T12:00:00Z"
```
Comment (marians) [inline aws/v12.1.1/release.yaml:62]: TODO: Update once more when we know when to release

Comment (calvix) [reply aws/v12.1.1/release.yaml:62]: true

**Comment by calvix** [comment]
I found some error when upgrading v12.1.0 to v12.1.1 unrelated to my changes and it can cause issues.

The chart-operator fails to install new cert-manager after the update due this error:
```
{"caller":"github.com/giantswarm/backoff@v0.2.0/notifier.go:13","controller":"chart-operator-chart","event":"update","function":"ApplyCreateChange","level":"warning","loop":"9","message":"retrying backoff in '5s' due to error","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"release","stack":{"annotation":"rpc error: code = Unknown desc = networkpolicies.networking.k8s.io \"cert-manager-crd-install\" already exists","kind":"unknown","stack":[{"file":"/go/pkg/mod/github.com/giantswarm/helmclient@v0.2.4/helmclient.go","line":382}]},"time":"2020-08-14T15:18:51.019588+00:00","version":"2321"}


```

whole log
```
{"caller":"github.com/giantswarm/chart-operator/service/controller/chart/resource/release/create.go:33","controller":"chart-operator-chart","event":"update","function":"ApplyCreateChange","level":"debug","loop":"9","message":"creating release `cert-manager`","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"release","time":"2020-08-14T15:18:50.466826+00:00","version":"2321"}
{"caller":"github.com/giantswarm/chart-operator/service/controller/chart/resource/release/resource.go:107","controller":"chart-operator-chart","event":"update","function":"ApplyCreateChange","level":"debug","loop":"9","message":"patching annotations for chart CR `cert-manager` in namespace `giantswarm`","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"release","time":"2020-08-14T15:18:50.466853+00:00","version":"2321"}
{"caller":"github.com/giantswarm/chart-operator/service/controller/chart/resource/release/resource.go:148","controller":"chart-operator-chart","event":"update","function":"ApplyCreateChange","level":"debug","loop":"9","message":"no need to patch annotations for chart CR `cert-manager` in namespace `giantswarm`","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"release","time":"2020-08-14T15:18:50.470437+00:00","version":"2321"}
{"caller":"github.com/giantswarm/helmclient@v0.2.4/helmclient.go:381","controller":"chart-operator-chart","event":"update","function":"ApplyCreateChange","level":"debug","loop":"9","message":"err string: `rpc error: code = Unknown desc = networkpolicies.networking.k8s.io \"cert-manager-crd-install\" already exists`","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"release","time":"2020-08-14T15:18:51.019403+00:00","version":"2321"}
{"caller":"github.com/giantswarm/backoff@v0.2.0/notifier.go:13","controller":"chart-operator-chart","event":"update","function":"ApplyCreateChange","level":"warning","loop":"9","message":"retrying backoff in '5s' due to error","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"release","stack":{"annotation":"rpc error: code = Unknown desc = networkpolicies.networking.k8s.io \"cert-manager-crd-install\" already exists","kind":"unknown","stack":[{"file":"/go/pkg/mod/github.com/giantswarm/helmclient@v0.2.4/helmclient.go","line":382}]},"time":"2020-08-14T15:18:51.019588+00:00","version":"2321"}
{"caller":"github.com/giantswarm/chart-operator/service/controller/chart/resource/release/create.go:100","controller":"chart-operator-chart","event":"update","function":"ApplyCreateChange","level":"debug","loop":"9","message":"release still being created","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"release","time":"2020-08-14T15:18:53.739003+00:00","version":"2321"}
{"caller":"github.com/giantswarm/chart-operator/service/controller/chart/resource/release/create.go:101","controller":"chart-operator-chart","event":"update","function":"ApplyCreateChange","level":"debug","loop":"9","message":"canceling resource","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"release","time":"2020-08-14T15:18:53.739053+00:00","version":"2321"}
{"caller":"github.com/giantswarm/chart-operator/service/controller/chart/resource/status/create.go:27","controller":"chart-operator-chart","event":"update","level":"debug","loop":"9","message":"getting status for release `cert-manager`","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"status","time":"2020-08-14T15:18:53.739192+00:00","version":"2321"}
{"caller":"github.com/giantswarm/chart-operator/service/controller/chart/resource/status/create.go:31","controller":"chart-operator-chart","event":"update","level":"debug","loop":"9","message":"release `cert-manager` not found","object":"/apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/cert-manager","resource":"status","time":"2020-08-14T15:18:53.810923+00:00","version":"2321"}
```


Can you take a look @giantswarm/team-halo-engineers?

**Comment by glitchcrab** [comment]
OK, I've spent some time looking into this and there are various things going on.

Firstly, the issue installing cert-manager is due to a bug in Helm 2 - the crd-install job (which runs before the chart is installed) sometimes fails. The crd-install resources _should_ be deleted by Tiller in the case of a failure, but Tiller sometimes fails to remove the resources. Then when the chart is applied again, the resources already exist with those names which produces the error seen above.

Secondly, had cert-manager deployed correctly then kiam still would not have worked. 12.1.0 didn't pull in v1.4.0 of kiam which was used during cert-manager testing - this version is needed because older versions of kiam create resources in the wrong API group (cert-manager's api group has changed). 

Do we have anyone using 12.1.0? If not then I don't think the upgrade test serves any purpose - kiam is broken in 12.1.0, and the cert-manager install issue isn't caused by the Chart - it's a Helm problem.

**Comment by glitchcrab** [comment]
@calvix @njuettner @alex-dabija i'd appreciate some thoughts here please: https://github.com/giantswarm/giantswarm/issues/12798

TLDR: i want to bump the version of cert-manager included in this release to work around the issue Vaclav mentioned above.

**Comment by calvix** [comment]
go ahead

**Comment by sslavic** [comment]
Included external-dns app upgrade v1.3.0

File: aws/v12.1.1/release.diff [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    version: 1.6.3                                                     version: 1.6.3
+  - name: aws-operator                                               - name: aws-operator
+    releaseOperatorDeploy: true                                        releaseOperatorDeploy: true
+    version: 8.7.5                                              |      version: 8.8.0
+  - name: calico                                                     - name: calico
+    version: 3.15.1                                                    version: 3.15.1
+  - name: cert-operator                                              - name: cert-operator
+    releaseOperatorDeploy: true                                        releaseOperatorDeploy: true
+    version: 0.1.0                                                     version: 0.1.0
+    reference: 0.1.0-2                                          <
```
Comment (sslavic) [inline aws/v12.1.1/release.diff:53]: I've updated the diff, and this change showed up. Is cert-operator reference intentionally removed in 12.1.1, or was it introduced retroactively in 12.1.0? If latter, should same or some different cert-operator reference be used in 12.1.1 as well?

Maybe @giantswarm/team-ludacris as cert-operator owners can help answer this one.

Comment (tfussell) [reply aws/v12.1.1/release.diff:53]: It must be `0.1.0-2` in all releases so please update it in `release.yaml` manually for now. I'm checking the latest `devctl` and it uses [apiextensions@0.4.19](https://github.com/giantswarm/apiextensions/blob/v0.4.19/pkg/apis/release/v1alpha1/release_types.go#L107) which includes this field so it should be preserved when reading, editing, and writing the new release. I'm not sure why it is being lost. I will try to investigate later today or tomorrow.

Comment (sslavic) [reply aws/v12.1.1/release.diff]: Done, thanks!

**Comment by sslavic** [comment]
I've tested on giraffe current release definition, 12.1.1 cluster creation, external-dns works.

Still, I've observed that the two app helm releases were in failed state, net-exporter and kiam:

```
{⎈ |giantswarm-fwh8b:default} ➜  ~ helm2 ls --tiller-namespace giantswarm                 
NAME                    	REVISION	UPDATED                 	STATUS  	CHART                             	APP VERSION	NAMESPACE   
cert-exporter           	1       	Tue Aug 18 22:33:59 2020	DEPLOYED	cert-exporter-1.2.3               	1.2.3      	kube-system 
chart-operator          	2       	Tue Aug 18 22:37:22 2020	DEPLOYED	chart-operator-0.13.2             	0.13.2     	giantswarm  
cluster-autoscaler      	1       	Tue Aug 18 22:34:05 2020	DEPLOYED	cluster-autoscaler-app-1.17.3     	v1.17.3    	kube-system 
coredns                 	1       	Tue Aug 18 22:37:23 2020	DEPLOYED	coredns-app-1.2.0                 	1.6.5      	kube-system 
etcd-cluster-migrator   	1       	Tue Aug 18 22:33:57 2020	DEPLOYED	etcd-cluster-migrator-1.0.2       	v1.0.2     	kube-system 
external-dns            	1       	Tue Aug 18 22:37:42 2020	DEPLOYED	external-dns-app-1.3.0            	v0.7.3     	kube-system 
kiam                    	1       	Tue Aug 18 22:37:44 2020	FAILED  	kiam-app-1.4.0                    	3.6        	kube-system 
kube-state-metrics      	1       	Tue Aug 18 22:33:58 2020	DEPLOYED	kube-state-metrics-app-1.1.1      	v1.9.7     	kube-system 
loadtest-app            	1       	Tue Aug 18 22:35:46 2020	DEPLOYED	loadtest-app-0.2.0                	0.0.1      	loadtest-app
metrics-server          	1       	Tue Aug 18 22:37:13 2020	DEPLOYED	metrics-server-app-1.1.1          	0.3.6      	kube-system 
net-exporter            	1       	Tue Aug 18 22:37:05 2020	FAILED  	net-exporter-1.9.0                	1.9.0      	kube-system 
nginx-ingress-controller	1       	Tue Aug 18 22:35:27 2020	DEPLOYED	nginx-ingress-controller-app-1.9.1	v0.34.1    	kube-system 
node-exporter           	1       	Tue Aug 18 22:34:05 2020	DEPLOYED	node-exporter-app-1.3.0           	v1.0.1     	kube-system 
```

Helm release history for the two showed a bit more information but I'm still puzzled what happened there:

```
{⎈ |giantswarm-fwh8b:default} ➜  ~ helm2 history --tiller-namespace giantswarm net-exporter -o yaml            
- chart: net-exporter-1.9.0
  description: 'Release "net-exporter" failed: Internal error occurred: failed to
    allocate a serviceIP: etcdserver: request timed out'
  revision: 1
  status: FAILED
  updated: Tue Aug 18 22:37:05 2020
```

```
{⎈ |giantswarm-fwh8b:default} ➜  ~ helm2 history --tiller-namespace giantswarm kiam -o yaml        
- chart: kiam-app-1.4.0
  description: 'Release "kiam" failed post-install: etcdserver: request timed out'
  revision: 1
  status: FAILED
  updated: Tue Aug 18 22:37:44 2020
```

Is it just China slowness side-effect, that can be tolerated?
Anyway, will leave test cluster up, fwh8b on giraffe, for further investigation in the morning if someone would like to have a look.

**Comment by marians** [comment]
@sslavic To keep the scope of this release as much focused as possible on the known problems to solve, I am going to revert the external-dns upgrade. The changes in external-dns are numerous, yet we don't know whether they fix a problem that occurred on our installations. Also we've been taking too long with this one here and don't want to spend any more time on testing.

**Comment by sslavic** [comment]
@marians I already did the testing, and at least with external-dns is concerned test was successful. I uncovered some other issues.

**Review by alex-dabija — APPROVED** [review]

**Review by njuettner — APPROVED** [review]
